### PR TITLE
Preliminary OCR implementation, ID/stashing logic overhaul, Belt/Consumibles overhaul

### PR DESCRIPTION
**Pull request by mgleed**
Hey! Really big PR here, sorry. :) Will notify when ready to review to merge, just have one or two more rounds of testing to go with some of the devs.

### Breaking changes:
-Nilhathak spelling fixed, so adjust your `[routes]` section in custom.ini accordingly
-`always_repair` config parameter changed to `runs_per_repair`. Set to 1 if you desire to always repair, set to 0 to only repair when the icon shows up, otherwise set to how many runs you want to do before doing a routine repair
-Added a line to graphics config, perform an `F9` on starting botty
-Rebuild conda environment is needed if you plan to run from source:
```
conda deactivate
conda env remove --name botty
conda env create environment.yml
conda activate botty
```

### Caveats:
-All drops and item description boxes are being read by OCR but not currently implemented in logic, this has led to an impact on performance during pickit routine. This is kept in order to help us generate truth to train the OCR. Once the OCR model is fully mature and pickit system overhauled, this should become a non-issue.

### TO-DO:
-Bugfixes. VM users seem to struggle most.
-Splitting most of the features into separate PR's--at least the ones that aren't intertwined.
-A2 Lysander doesn't sell TP scrolls but does sell pots, so may need to refactor town_manager.py

### Features:

**OCR**
-OCR model trained with Tesseract 5.0 using in-game and auto-generated D2 assets
-Adds ocr.py which contains functions for performing OCR using TesserOCR
-Limited amount of parsing leveraging the Parse module, currently implemented with checking TP, ID, and key counts from item descriptions
-Item parsing and pickit list generators are currently WIP and not yet implemented on the branch

**Item identification**
-Now supports tome of ID in inventory and will utilize if present, otherwise will utilize cain if `id_items` config parameter = 1
-No item filtering with OCR yet implemented, but this is planned

**Towning overhaul**
-Now checks items every pickup and decides to drop, ID, sell, or stash at the start of on_maintenance(), so no longer will go to the stash needlessly
-`always_repair` changed to `runs_per_repair` so you can set a more specific frequency for repairing if desired (see breaking changes above)
-Added a `sell_items` char config, set to 0 if you want to drop all items that won't be kept, to 1 to sell all items that won't be kept 
-When purchasing items, will use "shift" click to buy stacks if appropriate
-Consolidated purchasing functions to `transfer_items()` function in `inventory_manager.py`
-Reduced various redundant or excessively frequent actions with towning/shopping

**Consumibles**
-Split up the functions of `ui_manager.py` and `belt_manager.py` into three new classes: UIManager, InventoryManager, and ConsumiblesManager
-TP, ID, and key counts are now tracked with potion counts 
-Will buy TP, ID, and keys from vendors as needed

**Pickit**
-Will now pickup TP scrolls and ID scrolls if these are needed and if you have these set to 1 in the pickit.ini config file.

**Discord**
-Will send picture of every kept drop as well as the read OCR text 

**Other**
-Changed item spacing width to controller spacing for better item segmentation
-Added more routes to act 5
-Adjusted logic for misc utility `color_filter` to allow H to span over 0 or 180 for reds 
-Added support for list of images to be used with template_finder.search
-Added a couple more misc utility functions `mask_by_roi` (used to generate masks from img+roi), `trim_black` (used to trim black borders from image)

### CODE REVIEW

**Comment by mgleed** [comment]
If anyone would like to help with this branch, I could still use your help with OCR error detection. Run the game as usual, but make sure you have addressed the issues in ***Breaking Changes*** above and have these variables set as follows in your `custom.ini` file:
```ini
[general]
logg_lvl=debug
loot_screenshots=1

[char]
id_items=1
```

Once you have run long enough, send me a zip of the `loot_screenshots` directory and your `info.log` file.

**Comment by freesoul** [comment]
I am just wondering, what is the accuracy rate on the tesseract results? I trained a model with D2 fonts as well, and I get for "Harrogath" lots of variatns like "Hawogaln" or such

**Comment by mgleed** [comment]
> I am just wondering, what is the accuracy rate on the tesseract results? I trained a model with D2 fonts as well, and I get for "Harrogath" lots of variatns like "Hawogaln" or such

Right now it's the eng model refined for a single d2 font and only in the setting of item drops and item description boxes. About 99.5% word accuracy here for drops and item properties.

File: src/game_controller.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def run_bot(self, pick_corpse: bool = False):
         # Register that thread to the death and health manager so they can stop the bot thread if needed
         self.death_manager.set_callback(lambda: self.bot.stop() or kill_thread(self.bot_thread))
         self.health_manager.set_callback(lambda: self.bot.stop() or kill_thread(self.bot_thread))
-        self.health_manager.set_belt_manager(self.bot.get_belt_manager())
+        self.health_manager.set_consumibles_manager(self.bot.get_consumibles_manager())
```
Comment (VladimirMakaev) [inline src/game_controller.py:58]: typo here: "consumables"

File: src/health_manager.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 from template_finder import TemplateFinder
 from ui import UiManager
-from ui import BeltManager
+from ui import ConsumiblesManager
```
Comment (VladimirMakaev) [inline src/health_manager.py:3]: ConsumablesManager

File: src/health_manager.py [hunk 3]
```diff
@@ -20,7 +20,7 @@ def __init__(self, screen: Screen, template_finder: TemplateFinder):
         self._screen = screen
         self._template_finder = template_finder
         self._ui_manager = UiManager(screen, self._template_finder)
-        self._belt_manager = None # must be set with the belt manager form bot.py
+        self._consumibles_manager = None # must be set with the belt manager from bot.py
```
Comment (VladimirMakaev) [inline src/health_manager.py:23]: nit

File: src/health_manager.py [hunk 4]
```diff
@@ -34,8 +34,8 @@ def __init__(self, screen: Screen, template_finder: TemplateFinder):
     def stop_monitor(self):
         self._do_monitor = False
 
-    def set_belt_manager(self, belt_manager: BeltManager):
-        self._belt_manager = belt_manager
+    def set_consumibles_manager(self, consumibles_manager: ConsumiblesManager):
+        self._consumibles_manager = consumibles_manager
```
Comment (VladimirMakaev) [inline src/health_manager.py:38]: nit

**Comment by VladimirMakaev** [comment]
This is tremendous effort :) and can enable botty to be super smart :) From my point of view one I would try to split this PR in several to minimize the risk of regression and to simplify the review. E.g. 1) Nihlathak typo. 2) OCR as Framework 3) Splitting UI Manager. 4) Integrating OCR in XXX, etc.

**Comment by mgleed** [comment]
> This is tremendous effort :) and can enable botty to be super smart :) From my point of view one I would try to split this PR in several to minimize the risk of regression and to simplify the review. E.g. 1) Nihlathak typo. 2) OCR as Framework 3) Splitting UI Manager. 4) Integrating OCR in XXX, etc.

Yes, working on splitting it up. Not really ready for review yet anyway as it's very much in draft form. Currently put in the nihlathak & some misc util changes in separate PR's now that they've been validated to work in this branch; others to come. Will wait on aeon's "epic refactoring" and then adapt my changes to ui_manager and belt_manager. Will probably make an independent OCR function soon, currently working on a universal D2 model that works beyond the item drops and description boxes.

File: src/health_manager.py (hunk 2 above)
Comment (mgleed) [reply src/health_manager.py:3]: Oh lord you're right lol

**Comment by mgleed** [comment]
Not actively working on this branch due to #489 , will make small PR's for some of the features in the coming weeks.

**Pull request closed without merging**